src/ZipkinBundle/Components/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+namespace ZipkinBundle\Components\HttpClient;
+
+use Throwable;
+use Zipkin\Tracer;
+use Zipkin\Tracing;
+use Zipkin\SpanCustomizer;
+use Zipkin\Propagation\Map;
+use const Zipkin\Tags\ERROR;
+use Zipkin\SpanCustomizerShield;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
+final class HttpClient implements HttpClientInterface
+{
+    /**
+     * @var HttpClientInterface
+     */
+    private $delegate;
+
+    /**
+     * @var Tracer
+     */
+    private $tracer;
+
+    /**
+     * @var callable
+     */
+    private $injector;
+
+    /**
+     * @var HttpParser
+     */
+    private $httpParser;
+
+    public function __construct(
+        HttpClientInterface $client,
+        Tracing $tracing,
+        HttpParser $httpParser = null
+    ) {
+        $this->delegate = $client;
+        $this->tracer = $tracing->getTracer();
+        $this->injector = $tracing->getPropagation()->getInjector(new Map());
+        $this->httpParser = $httpParser ?? new DefaultHttpParser();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $span = $this->tracer->nextSpan();
+        $span->start();
+
+        $spanCustomizer = new SpanCustomizerShield($span);
+        $this->httpParser->request(strtolower($method), $url, $options, $spanCustomizer);
+
+        $headers = [];
+        ($this->injector)($span->getContext(), $headers);
+
+        try {
+            $options['headers'] = $headers + ($options['headers'] ?? []);
+            $options['on_progress'] = self::buildOnProgress(
+                $options['on_progress'] ?? null,
+                [$this->httpParser, 'response'],
+                $spanCustomizer,
+                [$span, 'finish']
+            );
+
+            // Since the cancel event is not being catched by the on_progress
```

Comment:
this maps to https://github.com/orgs/symfony/projects/1#card-30505178
but it doesn't happen today, and might never happen actually.

Reply:
Thanks for pointing it out. Then, I think my best option is to wrap the response and wait for the `cancel` to be called as of now.